### ISSUE TITLE
[Track] Detection on tracking

### ISSUE THREAD

**Issue by liunutt** [ultralytics/ultralytics#21163]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I have tested object tracking using YOLO segmentation, but I encountered a few issues:

1. I would like to configure the tracking YAML file to save the segmentation results without performing tracking.
_Tracking_
`2: 1216x1600 1 Guardrail`
_Without  Tracking_
`3: 1216x1600 1 Guardrail, 4 GuidePosts, 4 HighMasts, 2 InformationSignss, 2 MandatorySignss, 1 PedestrianBridge, 2 WarningSigns`

2. I want to adjust the bounding box to tightly fit around the segmentation mask, as shown in the example below.

![Image](https://github.com/user-attachments/assets/06d2f6d8-0a24-47fd-9c9c-ed431ccd11e7)

FYI: I have tested with yolo11 and BoTSort.

```yaml
tracker_type: botsort 

track_high_thresh: 0.25 
track_low_thresh: 0.1
new_track_thresh: 0.25 
track_buffer: 30 
match_thresh: 0.8 
fuse_score: True 
min_box_area: 0


gmc_method: sparseOptFlow # method of global motion compensation

proximity_thresh: 0.5
appearance_thresh: 0.25
with_reid: False
```

Could you recommend me how to fix these problems?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @liunutt, thank you for your interest in Ultralytics 🚀! This is an automated response to help guide you while an Ultralytics engineer reviews your issue and will assist you soon.

We recommend checking the [Docs](https://docs.ultralytics.com/) for detailed guides on tracking and segmentation, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many common configuration and results questions are already covered there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community wherever you feel most comfortable:
- For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions, check out [Discourse](https://community.ultralytics.com/)
- Or explore threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community

## Upgrade

Make sure you are using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). Upgrade with:

```bash
pip install -U ultralytics
```

## Environments

YOLO models can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your detailed description and configuration snippet! An Ultralytics engineer will review your questions regarding segmentation result saving and bounding box adjustment with YOLO11 and BoTSort shortly.

**Comment by Y-T-G** [comment]
1. Tracking removes any object with no track IDs. Each object needs to appear two consecutive frames for them to get track ID.
2. The box shift is probably due to Kalman filter smoothing. You will have to disable Kalman filter to remove the shift.
3. If all the objects are unconfirmed, they will not have track IDs.

**Comment by Y-T-G** [comment]
You can also use this PR with `min_hits` set to 1 to disable the detection dropping behavior.

https://github.com/ultralytics/ultralytics/pull/18108

**Comment by liunutt** [comment]
1. In video testing, there are more than 2 consecutive objects. Does it should to `PR` to setting `min_hints`?
2. How to adjust Kalman filter?

**Comment by glenn-jocher** [comment]
For your follow-up questions:

1. Yes, you can try the PR #18108 approach by setting `min_hits: 1` in your tracker YAML config - this should help retain more detections that would otherwise be dropped.

2. The Kalman filter smoothing is built into the tracking algorithms and isn't directly configurable through the YAML parameters. If you need tighter bounding boxes around segmentation masks, you might need to modify the tracker code directly or consider using a different approach where you run segmentation inference without tracking (`model.predict()`) when you need the raw detection results, and use tracking (`model.track()`) only when you need object IDs.

Regarding the detection count difference between tracking and non-tracking modes, this is expected behavior since tracking applies additional filtering to maintain consistent object identities across frames.

**Comment by liunutt** [comment]
@glenn-jocher Thank you for your support. I have some questions in the PR `min_hint` can solve small/tiny object track?

**Comment by liunutt** [comment]
Despite trying the min_hits parameter from [#18108](https://github.com/ultralytics/ultralytics/issues/18108), it still doesn't improve tracking for small objects in my case. Interestingly, the model can detect the small object correctly, but the tracker fails to maintain its identity across frames.

Here’s the test code I’m using for tracking:
```python
# Tracker arguments
args = SimpleNamespace(
    tracker_type="botsort",
    track_high_thresh=0.25,
    track_low_thresh=0.1,
    new_track_thresh=0.25,
    track_buffer=30,
    match_thresh=0.8,
    fuse_score=True,
    gmc_method="sparseOptFlow",
    proximity_thresh=0.5,
    appearance_thresh=0.25,
    with_reid=False,
    min_hits=1,  # Tried setting this to 1 for faster confirmation
)

# Initialize tracker
tracker = BOTSORT(args)
# For comparison, BYTETracker can be used instead:
# tracker = BYTETracker(args)

print("Running Ultralytics YOLO tracking...")
results = model(source=images)

tracked_objects = []

for idx, result in enumerate(results):
    frame = result.orig_img.copy()

    # Convert detection boxes to NumPy array
    det = result.boxes.cpu().numpy()

    if len(det) == 0:
        continue

    # Update tracker with current frame detections
    objects_in_this_frame = tracker.update(det, frame)
```

This is a sample frame from the test set.
above: detection , below: tracking

![Image](https://github.com/user-attachments/assets/58a0e8b5-9819-4950-b360-582d3cd024e7)

![Image](https://github.com/user-attachments/assets/41b02a39-7a5f-4dc1-bbd1-0e2cddc207ed)

**Comment by glenn-jocher** [comment]
For small object tracking issues, try lowering your confidence thresholds since small objects often have lower detection confidence scores. In your configuration, set `track_high_thresh: 0.1` and `new_track_thresh: 0.1` - this should help the tracker initialize and maintain tracks for smaller objects.

Also, consider using the standard YOLO tracking interface instead of directly calling the tracker, as it handles the detection-to-tracker format conversion more robustly:

```python
results = model.track(source=images, tracker='your_config.yaml', persist=True)
```

Small objects are inherently challenging for tracking due to their limited feature information and susceptibility to noise, so you may also want to experiment with `match_thresh: 0.6` (lower value = more lenient matching) and ensure your detection model is well-tuned for small objects in your specific use case.

**Comment by liunutt** [comment]
The detection model performs well, with confidence scores reaching up to 0.8 for the guide post (you can see the picture above). However, the tracking does not appear to work.

**Comment by glenn-jocher** [comment]
Since your detection confidence is good (0.8), the issue likely lies in the matching criteria being too strict for small objects. Try these parameter adjustments specifically for small object tracking:

```yaml
proximity_thresh: 0.3  # Lower IoU threshold for small objects
match_thresh: 0.5      # More lenient matching
track_buffer: 50       # Longer buffer to handle brief occlusions
gmc_method: ecc        # Often works better for small objects
```

Also, I'd recommend using the standard YOLO tracking interface instead of directly calling the tracker, as it handles edge cases better:

```python
results = model.track(source=images, tracker='your_config.yaml', persist=True, conf=0.25)
```

Small objects can be lost due to minimal pixel displacement between frames or when the tracker's motion model doesn't match their movement patterns. The parameter adjustments above should help with these common small object tracking challenges.
